src/config.zig

```diff
@@ -102,8 +96,15 @@ pub const pipelining_max = clients_max;
 pub const connection_delay_min_ms = 50;
 pub const connection_delay_max_ms = 1000;
 
-/// The maximum number of outgoing messages that may be queued on a connection.
-pub const connection_send_queue_max = pipelining_max;
+/// The maximum number of outgoing messages that may be queued on a replica connection.
+pub const connection_send_queue_max_replica = pipelining_max;
```

Comment:
> A better value for `send_queue_max` may be the number of requests from clients that the primary replica expects to receive all in the very same instant that it then needs to push to the send queue of the next replica, plus space for a control plane message, so perhaps `max(min(clients_max, 4), 2)`?

I don't understand the `max(min(..)..)` recommended value; can you elaborate? Based on your description, wouldn't it just be something like `clients_max + 1`?

---

Also, why not set the `tcp_snd_buf` to `connection_send_queue_max * message_size_max`?

Reply:
> Also, why not set the tcp_snd_buf to connection_send_queue_max * message_size_max?

The reason for not setting `tcp_snd_buf` too far beyond 4 MiB is an interesting production incident that Cloudflare ran into:

```
/// At the same time, be careful going beyond 4 MiB as the kernel may merge many small TCP packets,
/// causing considerable latency spikes for large buffer sizes:
/// https://blog.cloudflare.com/the-story-of-one-latency-spike/
```

> I don't understand the max(min(..)..) recommended value; can you elaborate? Based on your description, wouldn't it just be something like clients_max + 1?

`clients_max + 1` would be if all clients simultaneously sent a request over the network that arrive all in a single flush of our io_uring completion queue, faster than our memcopies to the kernel are able to empty the send_queue from user space to kernel space. This is an edge scenario. In practice, there would really be an ms or two between client request arrivals (i.e. multiple flushes of the completion queue), and our copies to kernel space when calling `send()` should be able to empty the send queue much faster than that rate. So the `max(min())` above (which would work out to around 2-4 messages in the send queue) should be sufficient.

It comes down to getting a good estimate on how many `recv()` syscall results we expect to flush from the completion queue in a single call to `io_uring_enter()`. Do we expect this to be between 25-50% of clients_max in the 99% case? Note also that io_uring will wake us the instant that we receive a single `recv()` syscall, and this already gives us a chance to do the corresponding memcopy through the send queue back to kernel space when replicating the prepare.

Clients will retry if the edge scenario happens, so there's no static allocation issue, only a latency spike issue, but it's very rare. We need to balance it against the memory cost of `clients_max` messages for all `replicas_max` connections, which is significant.

Reply:
Thank you, that makes sense.

I did read the cloudflare article, but that's not quite what I meant in my question. 

As I understand it, the `send_queue_max` is also the maximum number of messages that will go in the tcp send buffer. And since messages are (at most) 1M, that comes out to 4M when the `send_queue_max` is at its (new) maximum value of `4`. Isn't `send_queue_max * message_size_max` precisely the right value for `tcp_snd_buf`? Any less and the send queue might not be able to get all its data into the kernel. And more and it runs into the issue described in that blog post.

Reply:
Thanks for clarifying the motivation for setting `tcp_snd_buf ` to `send_queue_max * message_size_max`.

Let's do that, now that we have a tighter limit on the send queue. It does seem like we're approaching the new maximum of `4` from both directions this way. We could also double-check this against the typical NIC bandwidth we want to support. For example, how much snd buffer do we need to saturate a 1 Gbps or 10 Gbps NIC?